src/Simulation/QCTraceSimulator/SortedQubitPool.cs

```diff
@@ -0,0 +1,321 @@
+            internal ComplexTime Time; // This can be readonly if Sample handling is changed.
+
+            /// <summary>
+            /// Multiple nodes can correspond to the same time. They are added to the linked list via this field.
+            /// If qubits with the same time is common array may be more efficient than the linked list.
+            /// </summary>
+            internal QubitTimeNode NextNode;
+
+            /// <summary>
+            /// This is only used for Sample.
+            /// </summary>
+            internal QubitTimeNode() {
+            }
+
+            internal QubitTimeNode(long qubitId, ComplexTime time) {
+                QubitId = qubitId;
+                Time = time;
+                NextNode = null;
+            }
+
+            public override string ToString() {
+                return $"{QubitId} @ {Time}";
+            }
+        }
+
+        /// <summary>
+        /// Compares two QubitTimeNode objects for SortedSet.
+        /// It also collects minimum and maximum Time value of nodes been compared to the
+        /// sample Time since ResetForComparison is called.
+        /// </summary>
+        private class VisitingComparer : IComparer<QubitTimeNode>
+        {
+            /// <summary>
+            /// Maximum of all values, that are less than or equal than sample, among values seen after reset.
+            /// </summary>
+            internal ComplexTime MaxLowerBound = ComplexTime.MinValue;
+
+            /// <summary>
+            /// Minimum of all values, that are greater or equal than sample, among values seen after reset.
+            /// </summary>
+            internal ComplexTime MinUpperBound = ComplexTime.MaxValue;
+
+            /// <summary>
+            /// Sample, that will be passed to a series of comparisons.
+            /// Sample is not an object that's stored in the set so it is excluded from the values seen.
+            /// </summary>
+            private QubitTimeNode Sample;
+
+            /// <summary>
+            /// Resets minimum and maximum value of nodes seen from the set.
+            /// Also sets sample object, which should not be counted because it is not an object from the set.
+            /// </summary>
+            /// <param name="sample">Sample object that is not considered for minimum and maximum.</param>
+            internal void ResetForComparison(QubitTimeNode sample) {
+                MaxLowerBound = ComplexTime.MinValue;
+                MinUpperBound = ComplexTime.MaxValue;
+                Sample = sample;
+            }
+
+            /// <summary>
+            /// Compares Time field of two nodes. Also updates MaxLowerBound and MinUpperBound.
+            /// </summary>
+            /// <returns>Result of comparison of Time field of QubitTimeNode</returns>
+            public int Compare(QubitTimeNode a, QubitTimeNode b) {
+                // Note that comparison of a and b nodes to Sample is "by reference" in this function.
```

Comment:
I don't understand this part. Below, the Sample is allocated once with the default "0" qubit id and the code keeps overwriting its time when needed. How can this "by reference" comparison ever return 'true'?

Reply:
Comparison will return 'true' when one of the arguments of this compare is that same Sample instance allocated below. If it's another instance (even with the same values of fields) comparison by value will return false.